### PR TITLE
NoOp factory

### PR DESCRIPTION
**Pull request by xgouchet**
### What does this PR do?

Adds an annotation processing tool to automatically generate all the NoOp implementation of interfaces we have.
 - Automatically updates
 - Writes itself at compile time

### Motivation

Add nothing else to do this weekend 😄 

### Additional Notes

Anything else we should know when reviewing?

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)

### CODE REVIEW

File: tools/noopfactory/src/main/kotlin/com/datadog/tools/noopfactory/NoOpFactoryAnnotationProcessor.kt [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+            val paramType = param.asType().asKotlinTypeName()
+            val isNullable = param.getAnnotation(Nullable::class.java) != null
+            funSpec.addParameter(
+                param.simpleName.toString(),
+                paramType.copy(isNullable)
+            )
+        }
+
+        val returnType = element.returnType
+        val isReturnNullable = element.getAnnotation(Nullable::class.java) != null
```
Comment (mariusc83) [inline tools/noopfactory/src/main/kotlin/com/datadog/tools/noopfactory/NoOpFactoryAnnotationProcessor.kt:229]: you can move this inside the `if` statement

File: tools/noopfactory/src/main/kotlin/com/datadog/tools/noopfactory/NoOpFactoryAnnotationProcessor.kt [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+                typeSpec.addFunction(generateFunSpec(it))
+            }
+        }
+    }
+
+    /**
+     * Generates the implementation for a given method.
+     */
+    private fun generateFunSpec(element: ExecutableElement): FunSpec {
+        val funSpec = FunSpec.builder(element.simpleName.toString())
```
Comment (mariusc83) [inline tools/noopfactory/src/main/kotlin/com/datadog/tools/noopfactory/NoOpFactoryAnnotationProcessor.kt:214]: can you name this *funSpecBuilder* please. It is a bit confusing now.

File: tools/noopfactory/src/main/kotlin/com/datadog/tools/noopfactory/NoOpFactoryAnnotationProcessor.kt [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+     *  - if the return type is a primitive, the method will return 0/false
+     *  - if the return type is a String, the method will return an empty String
+     *  - if the return type is an enume, the method will return the first enum constant
+     *  - if the return type is an interface, it will assume a NoOp implementation exist
+     *  - otherwise it will assume a default constructor for the given type exists.
+     */
+    private fun generateFunReturnStatement(
+        returnType: TypeMirror,
+        isReturnNullable: Boolean,
+        funSpec: FunSpec.Builder
```
Comment (mariusc83) [inline tools/noopfactory/src/main/kotlin/com/datadog/tools/noopfactory/NoOpFactoryAnnotationProcessor.kt:251]: same here...name this funSpecBuilder please.

File: tools/noopfactory/src/main/kotlin/com/datadog/tools/noopfactory/TypeMirrorToKotlinTypeNameVisitor.kt [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+import javax.lang.model.type.TypeKind
+import javax.lang.model.type.TypeMirror
+import javax.lang.model.type.TypeVariable
+import javax.lang.model.type.WildcardType
+import javax.lang.model.util.SimpleTypeVisitor8
+
+/**
+ * Code mostly inspired by KotlinPoet, adapted to avoid some nullability issues
+ */
+internal class TypeMirrorToKotlinTypeNameVisitor : SimpleTypeVisitor8<TypeName, Void?>() {
```
Comment (mariusc83) [inline tools/noopfactory/src/main/kotlin/com/datadog/tools/noopfactory/TypeMirrorToKotlinTypeNameVisitor.kt:53]: Do you use any method in this class ?

File: tools/noopfactory/src/test/kotlin/com/datadog/tools/noopfactory/NoOpFactoryAnnotationProcessorTest.kt [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.CsvSource
+
+internal class NoOpFactoryAnnotationProcessorTest {
+
+    @ParameterizedTest
+    @CsvSource(
+        delimiter = ':',
+        value = [
```
Comment (mariusc83) [inline tools/noopfactory/src/test/kotlin/com/datadog/tools/noopfactory/NoOpFactoryAnnotationProcessorTest.kt:21]: maybe we should add also an UnitTest for the case where you have an Interface which inherits from a super one.

File: tools/noopfactory/src/main/kotlin/com/datadog/tools/noopfactory/TypeMirrorToKotlinTypeNameVisitor.kt (hunk 4 above)
Comment (xgouchet) [reply tools/noopfactory/src/main/kotlin/com/datadog/tools/noopfactory/TypeMirrorToKotlinTypeNameVisitor.kt:53]: Of course, all of them. This is a visitor pattern, so I don't call them directly, it's used in `TypeExt.kt` :
 
```
val typeName = accept(TypeMirrorToKotlinTypeNameVisitor(), null)
```

**Pull request merged**